python/ql/src/semmle/python/dataflow/new/internal/DataFlowPrivate.qll

```
@@ -1008,6 +1028,207 @@ predicate subscriptReadStep(CfgNode nodeFrom, Content c, CfgNode nodeTo) {
+ * directly to `receiver`.
+ *
+ * The strategy is then via several read-, store-, and flow steps:
+ * 1. [Flow] Content is transferred from `iterable` to `TIterableSequence(sequence)` via a
+ *    flow step. From here, everything happens on the LHS.
+ *
+ * 1. [Flow] Content is transferred from `TIterableSequence(sequence)` to `sequence` via a
+ *    flow step.
+ *
+ * 1. [Read] Content is read from `TIterableSequence(sequence)` into  `TIterableElement(sequence)`.
+ *    If `sequence` is of type tuple, we will not read tuple content as that would allow
+ *    cross talk.
+ *
+ * 1. [Store] Content is stored from `TIterableElement(sequence)` to `sequence`.
+ *    Here the content type is chosen according to the type of sequence.
+ *
+ * 1. [Read] Content is read from `sequence` to its elements according to the type of `sequence`.
+ *    If the element is a plain variable, the target is the corresponding essa node.
+ *    If the element is itelf a sequence, with control-flow node `seq`, the target is `TIterableSequence(seq)`.
+ *    If the element is a starred variable, with control-flow node `v`, the target is `TIterableElement(v)`.
+ *
+ * 1. [Store] Content is stored from `TIterableElement(v)` to the essa variable for `v`, with
+ *    content type `ListElement`.
+ *
+ * 1. [Flow, Read, Store] The last 5 steps are repeated for all recursive elements which are sequences.
+ */
+module unpackinAssignment {
+  /** A direct (or top-level) target of an unpacking assignment */
+  class UnpackingAssignmentDirectTarget extends ControlFlowNode {
+    Expr value;
+
+    UnpackingAssignmentDirectTarget() {
+      this instanceof SequenceNode and
+      exists(Assign assign | this.getNode() = assign.getATarget() | value = assign.getValue())
+    }
+
+    Expr getValue() { result = value }
+  }
+
+  /** A (possibly recursive) target of an unpacking assignment */
+  class UnpackingAssignmentTarget extends ControlFlowNode {
```

Could this extend `SequenceNode` instead (in which case we can drop the cast in `getElement`)?

---

No, because not all targets are sequences. For instance in `a, (b, c) = iterable`, only `a, (b, c)` and (a, b) are sequence, while `a`, `b`, and `c` ar not.

---

Ah, good point. I was misled by the fact that the only place where it _isn't_ a `SequenceNode` is in the definition of `TIterableElement`.

Perhaps it would be better to split this out into two subclasses, one for the "leaves" and one for the internal sequence nodes? That way, we would also avoid the slight code smell of `getElement/1` silently failing at the leaves. And we could get rid of all of the `instanceof SequenceNode` guards.

(Or perhaps we don't even need the "leaves only" class, since we never actually do things with _just_ those nodes. A simple subclass that represents `UnpackingAssignmentSequenceTarget`s should suffice.)

---

`UnpackingAssignmentSequenceTarget` works nicely :-)